README.md
This project is the VPN node of the Let's Connect! software. It controls the
OpenVPN processes on the VPN server.


CHANGES.md
# Changelog

## 3.0.0 (...)
- remove `vpn-lib-common` dependency by adding the required files to the 
  project directly

## 2.0.1 (2019-04-26)
- update tests to deal with updates internal API error messages 
  (vpn-lib-common)

## 2.0.0 (2019-04-01)
- better error when number of vpnProtoPorts is not 1,2,4,8 or 16
- remove 2FA verification through OpenVPN
- when using "split tunnel" configuration, also set a static route to the VPN 
  server through the client's normal gateway to avoid problems when pushing 
  routes that contain the VPN server's public IP address
- remove compression framing support
- remove tls-auth support
- remove "multi instance" support
- update firewalling

## 1.1.2 (2018-11-09)
- add support for `blockLan` to block traffic to local LAN network when 
  connected to the VPN
- deal with `blockSmb` not necessarily being available as a configuration 
  option

## 1.1.1 (2018-10-21)
- support also 32 bit Fedora/CentOS for determining OpenVPN auth plugin path

## 1.1.0 (2018-10-15)
- drop OpenVPN 2.3 client support, requires >= 2.4 now
  - IPv6 default gateway routing fixes no longer pushed to clients
  - `AES-256-GCM` is required cipher now
- only use `auth SHA256` with `tls-auth`, not needed with `AES-256-GCM` and 
  `tls-crypt`

## 1.0.22 (2018-10-10)
- when DNS servers are set, but VPN is not used as default gateway _do_ send 
  the DNS addresses
- empty `dns` field, i.e. `[]` does no longer send any DNS servers, before it 
  would send the IPv4 and IPv6 gateway addresses to the VPN clients
- introduce `@GW4@` and `@GW6@` macros that can be used in `dns` that will be
  replaced by the IPv4 and IPv6 gateway addresses

## 1.0.21 (2018-10-05)
- implement detector for `auth-script-openvpn` plugin

## 1.0.20 (2018-09-19)
- when setting `tlsProtection` to `false`, no longer allow `AES-256-CBC` cipher 
  and do not push IPv6 route fix

## 1.0.19 (2018-09-10)
- cleanup autoloader so Psalm will be able to verify the scripts in and bin and
  libexec folder
- additional Psalm fixes
- automatically provide IPv4 / IPv6 gateway address as DNS servers when none
  specified, i.e. `dns` is `[]` and `defaultGateway` is `true`
- no longer use `DNS6`, use `DNS` also for IPv6 DNS servers
- always generate new server certificates when running `server-config` script

## 1.0.18 (2018-08-05)
- many `vimeo/psalm` fixes

## 1.0.17 (2018-07-02)
- now always issue an `/112` to OpenVPN processes no matter who many IP space
  is available (issue #23)

## 1.0.16 (2018-06-12)
- fix IPv6 prefix when there is only one VPN process (issue #22)

## 1.0.15 (2018-06-06)
- change the default `--auth-gen-token` lifetime to 12 hours when 2FA is used
- support `tlsProtection`, allows disabling `tls-auth` and `tls-crypt`

## 1.0.14 (2018-04-17)
- update default config to use `enableNat4` and `enableNat6` instead of 
  `useNat` to allow separate configuration of NAT for IPv4 and IPv6

## 1.0.13 (2018-04-12)
- no longer push `bypass-dhcp` to clients
- set `keepalive` to 25 seconds

## 1.0.12 (2018-04-05)
- autodetect authPlugin instead of requiring configuration option. If plugin is
  installed, use it

## 1.0.11 (2018-03-29)
- increase `keepalive` for UDP, remove it for TCP

## 1.0.10 (2018-03-15)
- firewall config template change, a port is an integer, not a string

## 1.0.9 (2018-02-25)
- remove hacks for supporting 2.3 clients when `tlsCrypt` is enabled

## 1.0.8 (2018-01-17)
- autodetect RHEL/CentOS/Fedora or Debian/Ubuntu, no longer need the `--debian` 
  flag for `vpn-server-node-generate-firewall`

## 1.0.7 (2017-12-17)
- cleanup autoloading

## 1.0.6 (2017-12-15)
- push `comp-lzo no` to client when compression is enabled to disable 
  "adaptive compression" in the client
- update `eduvpn/common`

## 1.0.5 (2017-11-20)
- support PHPUnit 6
- add `certificate-info` script to show when the OpenVPN server certificates
  will expire
- restructure server configuration file generation
- Psalm fixes
- no longer push `comp-lzo no`, not needed as we don't actually use compression
- use same IPv6 default gateway routes on 2.3 clients as are used for 2.4 
  clients
- add tests for testing server configuration generation
- support disabling compression

## 1.0.4 (2017-10-25)
- remove `--profile` option for generating server configuration, generate for
  all profiles by default

## 1.0.3 (2017-10-20)
- only push `explicit-exit-notify` when using UDP
- support for "auth-script-openvpn" plugin for more efficient 2FA integration

## 1.0.2 (2017-09-29)
- expire 2FA connections after 8 hours, i.e. require new OTP code (#15)

## 1.0.1 (2017-07-28)
- allow specifying source IP range(s) for INPUT packet filter (#13)

## 1.0.0 (2017-07-13)
- initial release


src/OpenVpn.php
<?php

/*
 * eduVPN - End-user friendly VPN.
 *
 * Copyright: 2016-2019, The Commons Conservancy eduVPN Programme
 * SPDX-License-Identifier: AGPL-3.0+
 */

namespace LC\Node;

use DateTime;
use DateTimeZone;
use LC\Node\HttpClient\ServerClient;
use RuntimeException;

class OpenVpn
{
    // CentOS
    const LIBEXEC_DIR = '/usr/libexec/vpn-server-node';

    /** @var string */
    private $vpnConfigDir;

    /**
     * @param string $vpnConfigDir
     */
    public function __construct($vpnConfigDir)
    {
        FileIO::createDir($vpnConfigDir, 0700);
        $this->vpnConfigDir = $vpnConfigDir;
    }

    /**
     * @param string $vpnTlsDir
     * @param string $commonName
     *
     * @return void
     */
    public function generateKeys(ServerClient $serverClient, $vpnTlsDir, $commonName)
    {
        FileIO::createDir($vpnTlsDir, 0700);
        $certData = $serverClient->postRequireArray('add_server_certificate', ['common_name' => $commonName]);

        $certFileMapping = [
            'ca' => sprintf('%s/ca.crt', $vpnTlsDir),
            'certificate' => sprintf('%s/server.crt', $vpnTlsDir),
            'private_key' => sprintf('%s/server.key', $vpnTlsDir),
            'ta' => sprintf('%s/ta.key', $vpnTlsDir),
        ];

        foreach ($certFileMapping as $k => $v) {
            FileIO::writeFile($v, $certData[$k], 0600);
        }
    }

    /**
     * @param string $vpnUser
     * @param string $vpnGroup
     *
     * @return void
     */
    public function writeProfiles(ServerClient $serverClient, $vpnUser, $vpnGroup)
    {
        $profileList = $serverClient->getRequireArray('profile_list');

        $profileIdList = array_keys($profileList);
        foreach ($profileIdList as $profileId) {
            $profileConfigData = $profileList[$profileId];
            $profileConfigData['_user'] = $vpnUser;
            $profileConfigData['_group'] = $vpnGroup;
            $profileConfig = new ProfileConfig($profileConfigData);
            $this->writeProfile($profileId, $profileConfig);

            // generate a CN based on date and profile, instance
            $dateTime = new DateTime('now', new DateTimeZone('UTC'));
            $dateString = $dateTime->format('YmdHis');
            $cn = sprintf('%s.%s', $dateString, $profileId);
            $vpnTlsDir = sprintf('%s/tls/%s', $this->vpnConfigDir, $profileId);

            $this->generateKeys($serverClient, $vpnTlsDir, $cn);
        }
    }

    /**
     * @param string $profileId
     *
     * @return void
     */
    public function writeProfile($profileId, ProfileConfig $profileConfig)
    {
        $range = new IP($profileConfig->getItem('range'));
        $range6 = new IP($profileConfig->getItem('range6'));
        $processCount = \count($profileConfig->getItem('vpnProtoPorts'));

        $allowedProcessCount = [1, 2, 4, 8, 16];
        if (!\in_array($processCount, $allowedProcessCount, true)) {
            throw new RuntimeException('"vpnProtoPorts" must contain 1,2,4,8 or 16 entries');
        }
        $splitRange = $range->split($processCount);
        $splitRange6 = $range6->split($processCount);

        $managementIp = $profileConfig->getItem('managementIp');
        $profileNumber = $profileConfig->getItem('profileNumber');

        $processConfig = [
            'managementIp' => $managementIp,
        ];

        for ($i = 0; $i < $processCount; ++$i) {
            list($proto, $port) = self::getProtoPort($profileConfig->getItem('vpnProtoPorts'), $profileConfig->getItem('listen'))[$i];
            $processConfig['range'] = $splitRange[$i];
            $processConfig['range6'] = $splitRange6[$i];
            $processConfig['dev'] = sprintf('tun-%d-%d', $profileConfig->getItem('profileNumber'), $i);
            $processConfig['proto'] = $proto;
            $processConfig['port'] = $port;
            $processConfig['local'] = $profileConfig->getItem('listen');
            $processConfig['managementPort'] = 11940 + $this->toPort($profileNumber, $i);
            $processConfig['configName'] = sprintf(
                '%s-%d.conf',
                $profileId,
                $i
            );

            $this->writeProcess($profileId, $profileConfig, $processConfig);
        }
    }

    /**
     * @param string $listenAddress
     * @param string $proto
     *
     * @return string
     */
    private static function getFamilyProto($listenAddress, $proto)
    {
        $v6 = false !== strpos($listenAddress, ':');
        if ('udp' === $proto) {
            return $v6 ? 'udp6' : 'udp';
        }
        if ('tcp' === $proto) {
            return $v6 ? 'tcp6-server' : 'tcp-server';
        }

        throw new RuntimeException('only "tcp" and "udp" are supported as protocols');
    }

    /**
     * @param string $listenAddress
     *
     * @return array
     */
    private static function getProtoPort(array $vpnProcesses, $listenAddress)
    {
        $convertedPortProto = [];

        foreach ($vpnProcesses as $vpnProcess) {
            list($proto, $port) = explode('/', $vpnProcess);
            $convertedPortProto[] = [self::getFamilyProto($listenAddress, $proto), $port];
        }

        return $convertedPortProto;
    }

    /**
     * @param string $profileId
     *
     * @return void
     */
    private function writeProcess($profileId, ProfileConfig $profileConfig, array $processConfig)
    {
        $tlsDir = sprintf('tls/%s', $profileId);

        $rangeIp = new IP($processConfig['range']);
        $range6Ip = new IP($processConfig['range6']);

        // static options
        $serverConfig = [
            'verb 3',
            'dev-type tun',
            sprintf('user %s', $profileConfig->getItem('_user')),
            sprintf('group %s', $profileConfig->getItem('_group')),
            'topology subnet',
            'persist-key',
            'persist-tun',
            'remote-cert-tls client',
            'tls-version-min 1.2',
            'tls-cipher TLS-ECDHE-RSA-WITH-AES-256-GCM-SHA384',
            'dh none', // Only ECDHE
            'ncp-ciphers AES-256-GCM',  // only AES-256-GCM
            'cipher AES-256-GCM',       // only AES-256-GCM
            'auth none',
            sprintf('client-connect %s/client-connect', self::LIBEXEC_DIR),
            sprintf('client-disconnect %s/client-disconnect', self::LIBEXEC_DIR),
            sprintf('ca %s/ca.crt', $tlsDir),
            sprintf('cert %s/server.crt', $tlsDir),
            sprintf('key %s/server.key', $tlsDir),
            sprintf('server %s %s', $rangeIp->getNetwork(), $rangeIp->getNetmask()),
            sprintf('server-ipv6 %s', $range6Ip->getAddressPrefix()),
            sprintf('max-clients %d', $rangeIp->getNumberOfHosts() - 1),
            'script-security 2',
            sprintf('dev %s', $processConfig['dev']),
            sprintf('port %d', $processConfig['port']),
            sprintf('management %s %d', $processConfig['managementIp'], $processConfig['managementPort']),
            sprintf('setenv PROFILE_ID %s', $profileId),
            sprintf('proto %s', $processConfig['proto']),
            sprintf('local %s', $processConfig['local']),
        ];

        if (!$profileConfig->getItem('enableLog')) {
            $serverConfig[] = 'log /dev/null';
        }

        if ('tcp-server' === $processConfig['proto'] || 'tcp6-server' === $processConfig['proto']) {
            $serverConfig[] = 'tcp-nodelay';
        }

        if ('udp' === $processConfig['proto'] || 'udp6' === $processConfig['proto']) {
            // notify the clients to reconnect to the exact same OpenVPN process
            // when the OpenVPN process restarts...
            $serverConfig[] = 'keepalive 10 60';
            $serverConfig[] = 'explicit-exit-notify 1';
            // also ask the clients on UDP to tell us when they leave...
            // https://github.com/OpenVPN/openvpn/commit/422ecdac4a2738cd269361e048468d8b58793c4e
            $serverConfig[] = 'push "explicit-exit-notify 1"';
        }

        if ('tls-crypt' === $profileConfig->getItem('tlsProtection')) {
            $serverConfig[] = sprintf('tls-crypt %s/ta.key', $tlsDir);
        }

        // Routes
        $serverConfig = array_merge($serverConfig, self::getRoutes($profileConfig));

        // DNS
        $serverConfig = array_merge($serverConfig, self::getDns($rangeIp, $range6Ip, $profileConfig));

        // Client-to-client
        $serverConfig = array_merge($serverConfig, self::getClientToClient($profileConfig));

        sort($serverConfig, SORT_STRING);

        $serverConfig = array_merge(
            [
                '#',
                '# OpenVPN Server Configuration',
                '#',
                '# ******************************************',
                '# * THIS FILE IS GENERATED, DO NOT MODIFY! *',
                '# ******************************************',
                '#',
            ],
            $serverConfig
        );

        $configFile = sprintf('%s/%s', $this->vpnConfigDir, $processConfig['configName']);

        FileIO::writeFile($configFile, implode(PHP_EOL, $serverConfig), 0600);
    }

    /**
     * @return array
     */
    private static function getRoutes(ProfileConfig $profileConfig)
    {
        if ($profileConfig->getItem('defaultGateway')) {
            $redirectFlags = ['def1', 'ipv6'];
            if ($profileConfig->hasItem('blockLan') && $profileConfig->getItem('blockLan')) {
                $redirectFlags[] = 'block-local';
            }

            return [
                sprintf('push "redirect-gateway %s"', implode(' ', $redirectFlags)),
            ];
        }

        // Always set a route to the remote host through the client's default
        // gateway to avoid problems when the "split routes" pushed also
        // contain a range with the public IP address of the VPN server.
        // When connecting to a VPN server _over_ IPv6, OpenVPN takes care of
        // this all by itself by setting a /128 through the client's original
        // IPv6 gateway
        $routeConfig = [
            'push "route remote_host 255.255.255.255 net_gateway"',
        ];

        // there may be some routes specified, push those, and not the default
        foreach ($profileConfig->getSection('routes')->toArray() as $route) {
            $routeIp = new IP($route);
            if (6 === $routeIp->getFamily()) {
                // IPv6
                $routeConfig[] = sprintf('push "route-ipv6 %s"', $routeIp->getAddressPrefix());
            } else {
                // IPv4
                $routeConfig[] = sprintf('push "route %s %s"', $routeIp->getAddress(), $routeIp->getNetmask());
            }
        }

        return $routeConfig;
    }

    /**
     * @return array
     */
    private static function getDns(IP $rangeIp, IP $range6Ip, ProfileConfig $profileConfig)
    {
        $dnsEntries = [];
        if ($profileConfig->getItem('defaultGateway')) {
            // prevent DNS leakage on Windows when VPN is default gateway
            $dnsEntries[] = 'push "block-outside-dns"';
        }
        $dnsList = $profileConfig->getSection('dns')->toArray();
        foreach ($dnsList as $dnsAddress) {
            // replace the macros by IP addresses (LOCAL_DNS)
            if ('@GW4@' === $dnsAddress) {
                $dnsAddress = $rangeIp->getFirstHost();
            }
            if ('@GW6@' === $dnsAddress) {
                $dnsAddress = $range6Ip->getFirstHost();
            }
            $dnsEntries[] = sprintf('push "dhcp-option DNS %s"', $dnsAddress);
        }

        return $dnsEntries;
    }

    /**
     * @return array
     */
    private static function getClientToClient(ProfileConfig $profileConfig)
    {
        if (!$profileConfig->getItem('clientToClient')) {
            return [];
        }

        $rangeIp = new IP($profileConfig->getItem('range'));
        $range6Ip = new IP($profileConfig->getItem('range6'));

        return [
            'client-to-client',
            sprintf('push "route %s %s"', $rangeIp->getAddress(), $rangeIp->getNetmask()),
            sprintf('push "route-ipv6 %s"', $range6Ip->getAddressPrefix()),
        ];
    }

    /**
     * @param int $profileNumber
     * @param int $processNumber
     *
     * @return int
     */
    private function toPort($profileNumber, $processNumber)
    {
        // we have 2^16 - 11940 ports available for management ports, so let's
        // say we have 2^14 ports available to distribute over profiles and
        // processes, let's take 12 bits, so we have 64 profiles with each 64
        // processes...
        return ($profileNumber - 1 << 6) | $processNumber;
    }
}
